 ## Data Cleaning

In [1]:
#!pip install --upgrade openpyxl

In [2]:
#importing libraries and installing packages needed

import pandas as pd

C:\Users\hello\anaconda3\Lib\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


In [3]:
# Load the two worksheets
df1 = pd.read_excel(
    "challenge dataset.xlsx", 
    sheet_name='Candidat au don 2019 (avec age)', 
    dtype={
        'ÉLIGIBILITÉ AU DON.': str,
        'A-t-il_(elle)_déjà_donné_le_sang_': str
    })
df2 = pd.read_excel("challenge dataset.xlsx", sheet_name='Donneurs 2019')

#first five columns of 'Candidat au don 2019 (avec age)' worksheet
df1.head()

ID  Age Horodateur Niveau_scolaire Genre_  Taille_  Poids  \
0  DONOR_1842   24    46:44.9   Universitaire  Homme      NaN    NaN   
1  DONOR_1843   23    44:19.0   Universitaire  Homme      NaN    NaN   
2  DONOR_1844   41    46:54.3           Aucun  Homme      NaN    NaN   
3  DONOR_1845   24    29:29.7   Universitaire  Homme      NaN    NaN   
4  DONOR_1846   30    47:47.4      Secondaire  Homme      NaN    NaN   

  Situation_Matrimoniale_(SM)   Profession_ Arrondissement_de_résidence_  ...  \
0                 Célibataire  Etudiant (e)                     Douala 3  ...   
1                 Célibataire  Sans Emplois                        R A S  ...   
2                   Marié (e)     Tailleur                      Douala 1  ...   
3                 Célibataire      Etudiant                     Douala 5  ...   
4                 Célibataire     Militaire        Douala (Non précisé )  ...   

  Raison_de_non-eligibilité_totale__[Porteur(HIV,hbs,hcv)]  \
0                                                NaN         
1                                                Non         
2                                                Non         
3                                                NaN         
4                                                NaN         

  Raison_de_non-eligibilité_totale__[Opéré]  \
0                                       NaN   
1                                       Oui   
2                                       Non   
3                                       NaN   
4                                       NaN   

  Raison_de_non-eligibilité_totale__[Drepanocytaire]  \
0                                                NaN   
1                                                Non   
2                                                Non   
3                                                NaN   
4                                                NaN   

  Raison_de_non-eligibilité_totale__[Diabétique]  \
0                                            NaN   
1                                            Non   
2                                            Non   
3                                            NaN   
4                                            NaN   

  Raison_de_non-eligibilité_totale__[Hypertendus]  \
0                                             NaN   
1                                             Non   
2                                             Non   
3                                             NaN   
4                                             NaN   

  Raison_de_non-eligibilité_totale__[Asthmatiques]  \
0                                              NaN   
1                                              Non   
2                                              Non   
3                                              NaN   
4                                              NaN   

  Raison_de_non-eligibilité_totale__[Cardiaque]  \
0                                           NaN   
1                                           Non   
2                                           Non   
3                                           NaN   
4                                           NaN   

  Raison_de_non-eligibilité_totale__[Tatoué]  \
0                                        NaN   
1                                        Non   
2                                        Non   
3                                        NaN   
4                                        NaN   

  Raison_de_non-eligibilité_totale__[Scarifié] Si_autres_raison_préciser_  
0                                          NaN                        NaN  
1                                          Non                        NaN  
2                                          Non        Aucune information   
3                                          NaN                        NaN  
4                                          NaN                        NaN  

[5 rows x 40 columns]

In [4]:
#first five coumns of 'Donneur 2019' worksheet
df2.head()

Horodateur Sexe Age  Type de donation   \
0 2020-07-16 19:02:33.994    M   26                 F   
1 2020-07-16 19:05:51.521    M   24                 F   
2 2020-07-17 04:52:09.054    M   41                 F   
3 2020-07-17 04:56:03.200    M   20                 F   
4 2020-07-17 05:49:40.730    M   34                 F   

  Groupe Sanguin ABO / Rhesus                       Phenotype   
0                           B+  RH-2, RH+4, RH-3, RH+5, -kell1  
1                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
2                           A+  RH+2, RH+4, RH-3, RH+5, -kell1  
3                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
4                           O+  RH-2, RH+4, RH+3, RH+5, -kell1

Moving on with cleaning one dataframe at a time. 

---

### Cleaning 'Candidat au don 2019 (avec age)' Sheet 

In [5]:
df1

ID  Age Horodateur Niveau_scolaire Genre_  Taille_  Poids  \
0     DONOR_1842   24    46:44.9   Universitaire  Homme      NaN    NaN   
1     DONOR_1843   23    44:19.0   Universitaire  Homme      NaN    NaN   
2     DONOR_1844   41    46:54.3           Aucun  Homme      NaN    NaN   
3     DONOR_1845   24    29:29.7   Universitaire  Homme      NaN    NaN   
4     DONOR_1846   30    47:47.4      Secondaire  Homme      NaN    NaN   
...          ...  ...        ...             ...    ...      ...    ...   
1841  DONOR_3683   26    31:07.0      Secondaire  Homme      NaN    NaN   
1842  DONOR_3684   23    29:02.4      Secondaire  Homme      NaN    NaN   
1843  DONOR_3685   30    33:42.3   Universitaire  Homme      NaN    NaN   
1844  DONOR_3686   28    02:54.5     Pas Précisé  Femme      NaN   96.0   
1845  DONOR_3687   29    56:38.9     Pas Précisé  Homme      NaN    NaN   

     Situation_Matrimoniale_(SM)           Profession_  \
0                    Célibataire          Etudiant (e)   
1                    Célibataire          Sans Emplois   
2                      Marié (e)             Tailleur    
3                    Célibataire              Etudiant   
4                    Célibataire             Militaire   
...                          ...                   ...   
1841                 Célibataire                 Élève   
1842                 Célibataire                Élève    
1843                 Célibataire  Technicien superieur   
1844                 Célibataire           Estheticien   
1845                 Célibataire           Électricien   

     Arrondissement_de_résidence_  ...  \
0                        Douala 3  ...   
1                           R A S  ...   
2                        Douala 1  ...   
3                        Douala 5  ...   
4           Douala (Non précisé )  ...   
...                           ...  ...   
1841                     Douala 3  ...   
1842        Douala (Non précisé )  ...   
1843        Douala (Non précisé )  ...   
1844        Douala (Non précisé )  ...   
1845                     Douala 3  ...   

     Raison_de_non-eligibilité_totale__[Porteur(HIV,hbs,hcv)]  \
0                                                   NaN         
1                                                   Non         
2                                                   Non         
3                                                   NaN         
4                                                   NaN         
...                                                 ...         
1841                                                NaN         
1842                                                NaN         
1843                                                NaN         
1844                                                NaN         
1845                                                NaN         

     Raison_de_non-eligibilité_totale__[Opéré]  \
0                                          NaN   
1                                          Oui   
2                                          Non   
3                                          NaN   
4                                          NaN   
...                                        ...   
1841                                       NaN   
1842                                       NaN   
1843                                       NaN   
1844                                       NaN   
1845                                       NaN   

     Raison_de_non-eligibilité_totale__[Drepanocytaire]  \
0                                                   NaN   
1                                                   Non   
2                                                   Non   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
1841                                                NaN   
1842                                                NaN  

In [6]:
# Display the column names
print(df1.columns.tolist())

['ID', 'Age', 'Horodateur', 'Niveau_scolaire', 'Genre_', 'Taille_', 'Poids', 'Situation_Matrimoniale_(SM)', 'Profession_', 'Arrondissement_de_résidence_', 'Quartier_de_Résidence_', 'Nationalité_', 'Religion_', 'A-t-il_(elle)_déjà_donné_le_sang_', 'Si_oui_preciser_la_date_du_dernier_don._', 'Taux_d’hémoglobine_', 'ÉLIGIBILITÉ_AU_DON.', 'Raison_indisponibilité__[Est_sous_anti-biothérapie__]', 'Raison_indisponibilité__[Taux_d’hémoglobine_bas_]', 'Raison_indisponibilité__[date_de_dernier_Don_<_3_mois_]', 'Raison_indisponibilité__[IST_récente_(Exclu_VIH,_Hbs,_Hcv)]', 'Date_de_dernières_règles_(DDR)__', 'Raison_de_l’indisponibilité_de_la_femme_[La_DDR_est_mauvais_si_<14_jour_avant_le_don]', 'Raison_de_l’indisponibilité_de_la_femme_[Allaitement_]', 'Raison_de_l’indisponibilité_de_la_femme_[A_accoucher_ces_6_derniers_mois__]', 'Raison_de_l’indisponibilité_de_la_femme_[Interruption_de_grossesse__ces_06_derniers_mois]', 'Raison_de_l’indisponibilité_de_la_femme_[est_enceinte_]', 'Autre_raisons,__

In [7]:
df1.shape

(1846, 40)

In [8]:
# Count NaN values per column
nan_count = df1.isna().sum()
print(nan_count)

ID                                                                                              0
Age                                                                                             0
Horodateur                                                                                      0
Niveau_scolaire                                                                                 0
Genre_                                                                                          0
Taille_                                                                                      1786
Poids                                                                                        1742
Situation_Matrimoniale_(SM)                                                                     0
Profession_                                                                                     0
Arrondissement_de_résidence_                                                                    0
Quartier_de_Résidenc

Given that the number of rows in the datset is 1846 and some columns have 95.77% of missing data , we shall proceed to droping them 

In [9]:
# Drop rows with any NaN values
clean_df1 = df1.drop(columns=[ 'Taille_','Poids','Si_oui_preciser_la_date_du_dernier_don._','Raison_indisponibilité__[Est_sous_anti-biothérapie__]', 'Raison_indisponibilité__[Taux_d’hémoglobine_bas_]', 'Raison_indisponibilité__[date_de_dernier_Don_<_3_mois_]', 'Raison_indisponibilité__[IST_récente_(Exclu_VIH,_Hbs,_Hcv)]', 'Date_de_dernières_règles_(DDR)__', 'Raison_de_l’indisponibilité_de_la_femme_[La_DDR_est_mauvais_si_<14_jour_avant_le_don]', 'Raison_de_l’indisponibilité_de_la_femme_[Allaitement_]', 'Raison_de_l’indisponibilité_de_la_femme_[A_accoucher_ces_6_derniers_mois__]', 'Raison_de_l’indisponibilité_de_la_femme_[Interruption_de_grossesse__ces_06_derniers_mois]', 'Raison_de_l’indisponibilité_de_la_femme_[est_enceinte_]', 'Autre_raisons,__preciser_', 'Sélectionner_"ok"_pour_envoyer_', 'Raison_de_non-eligibilité_totale__[Antécédent_de_transfusion]', 'Raison_de_non-eligibilité_totale__[Porteur(HIV,hbs,hcv)]', 'Raison_de_non-eligibilité_totale__[Opéré]', 'Raison_de_non-eligibilité_totale__[Drepanocytaire]', 'Raison_de_non-eligibilité_totale__[Diabétique]', 'Raison_de_non-eligibilité_totale__[Hypertendus]', 'Raison_de_non-eligibilité_totale__[Asthmatiques]', 'Raison_de_non-eligibilité_totale__[Cardiaque]', 'Raison_de_non-eligibilité_totale__[Tatoué]', 'Raison_de_non-eligibilité_totale__[Scarifié]', 'Si_autres_raison_préciser_'])
clean_df1

ID  Age Horodateur Niveau_scolaire Genre_  \
0     DONOR_1842   24    46:44.9   Universitaire  Homme   
1     DONOR_1843   23    44:19.0   Universitaire  Homme   
2     DONOR_1844   41    46:54.3           Aucun  Homme   
3     DONOR_1845   24    29:29.7   Universitaire  Homme   
4     DONOR_1846   30    47:47.4      Secondaire  Homme   
...          ...  ...        ...             ...    ...   
1841  DONOR_3683   26    31:07.0      Secondaire  Homme   
1842  DONOR_3684   23    29:02.4      Secondaire  Homme   
1843  DONOR_3685   30    33:42.3   Universitaire  Homme   
1844  DONOR_3686   28    02:54.5     Pas Précisé  Femme   
1845  DONOR_3687   29    56:38.9     Pas Précisé  Homme   

     Situation_Matrimoniale_(SM)           Profession_  \
0                    Célibataire          Etudiant (e)   
1                    Célibataire          Sans Emplois   
2                      Marié (e)             Tailleur    
3                    Célibataire              Etudiant   
4                    Célibataire             Militaire   
...                          ...                   ...   
1841                 Célibataire                 Élève   
1842                 Célibataire                Élève    
1843                 Célibataire  Technicien superieur   
1844                 Célibataire           Estheticien   
1845                 Célibataire           Électricien   

     Arrondissement_de_résidence_ Quartier_de_Résidence_  Nationalité_  \
0                        Douala 3                  Yassa  Camerounaise   
1                           R A S                  R A S         R A S   
2                        Douala 1               NEW BELL  Camerounaise   
3                        Douala 5           Ange raphael  Camerounaise   
4           Douala (Non précisé )                  Beedi  Camerounaise   
...                           ...                    ...           ...   
1841                     Douala 3               Ndogbong  Camerounaise   
1842        Douala (Non précisé )                   Pk14  Camerounaise   
1843        Douala (Non précisé )      Cite des palmiers  Camerounaise   
1844        Douala (Non précisé )          Douala douala  Camerounaise   
1845                     Douala 3             PK16 NSAPE  Camerounaise   

                  Religion_ A-t-il_(elle)_déjà_donné_le_sang_  \
0     Chretien (Catholique)                               Non   
1     Chretien (Catholique)                               Oui   
2               Pas Précisé                               Non   
3                  Musulman                               Non   
4                Adventiste                               Oui   
...                     ...                               ...   
1841   Chrétien non précisé                               Non   
1842            Pas Précisé                               Oui   
1843               Musulman                               Oui   
1844            Pas Précisé                               Non   
1845  Chretien (Catholique)                               Non   

     Taux_d’hémoglobine_          ÉLIGIBILITÉ_AU_DON.  
0                    NaN                     Eligible  
1                   10.8  Définitivement non-eligible  
2                    NaN  Définitivement non-eligible  
3                     15                     Eligible  
4                    NaN                     Eligible  
...                  ...                          ...  
1841                 NaN                     Eligible  
1842                 NaN                     Eligible  
1843                13.4                     Eligible  
1844                  12                     Eligible  
1845                14.5                     Eligible  

[1846 rows x 14 columns]

In [10]:
print(clean_df1.dtypes)

ID                                   object
Age                                   int64
Horodateur                           object
Niveau_scolaire                      object
Genre_                               object
Situation_Matrimoniale_(SM)          object
Profession_                          object
Arrondissement_de_résidence_         object
Quartier_de_Résidence_               object
Nationalité_                         object
Religion_                            object
A-t-il_(elle)_déjà_donné_le_sang_    object
Taux_d’hémoglobine_                  object
ÉLIGIBILITÉ_AU_DON.                  object
dtype: object


In [11]:
# Change data types
clean_df1['Niveau_scolaire'] = clean_df1['Niveau_scolaire'].astype('category')
clean_df1['Genre_'] = clean_df1['Genre_'].astype('category')
clean_df1['Situation_Matrimoniale_(SM)'] = clean_df1['Situation_Matrimoniale_(SM)'].astype('category')
clean_df1['A-t-il_(elle)_déjà_donné_le_sang_'] = clean_df1['A-t-il_(elle)_déjà_donné_le_sang_'].astype('category')
clean_df1['ÉLIGIBILITÉ_AU_DON.'] = clean_df1['ÉLIGIBILITÉ_AU_DON.'].astype('category')

In [12]:
# Convert other columns to string
string_columns = ['ID', 'Horodateur', 'Profession_', 'Arrondissement_de_résidence_', 
                  'Quartier_de_Résidence_', 'Nationalité_', 'Religion_']
clean_df1[string_columns] = clean_df1[string_columns].astype(str)

In [13]:
# Clean and convert 'Taux_d’hémoglobine_' to float
# Remove any non-numeric characters (e.g., ' g/dL') and convert to float
clean_df1['Taux_d’hémoglobine_'] = clean_df1['Taux_d’hémoglobine_'].str.replace(' g/dL', '', regex=False)
clean_df1['Taux_d’hémoglobine_'] = pd.to_numeric(clean_df1['Taux_d’hémoglobine_'], errors='coerce')


In [41]:
print(clean_df1.dtypes)

ID                                     object
Age                                     int64
Horodateur                             object
Niveau_scolaire                      category
Genre_                                  int32
Situation_Matrimoniale_(SM)             int32
Profession_                            object
Arrondissement_de_résidence_           object
Quartier_de_Résidence_                 object
Nationalité_                           object
Religion_                              object
A-t-il_(elle)_déjà_donné_le_sang_       int32
Taux_d’hémoglobine_                   float64
ÉLIGIBILITÉ_AU_DON.                     int32
dtype: object


In [14]:
# Imputing with mean
#mean_value = clean_df1['Taux_d’hémoglobine_'].mean()
#clean_df1['Taux_d’hémoglobine_'] = clean_df1['Taux_d’hémoglobine_'].fillna(mean_value)

using the characteristics of similar observations to fill in missing values

In [15]:
from sklearn.impute import KNNImputer
imputer = KNNImputer(n_neighbors=5)
clean_df1[['Taux_d’hémoglobine_']] = imputer.fit_transform(clean_df1[['Taux_d’hémoglobine_']])

In [16]:
print(clean_df1['ÉLIGIBILITÉ_AU_DON.'].unique())

['Eligible', 'Définitivement non-eligible', 'Temporairement Non-eligible']
Categories (3, object): ['Définitivement non-eligible', 'Eligible', 'Temporairement Non-eligible']


In [17]:
#dftest = pd.read_excel("Updated Challenge dataset.xlsx", sheet_name='Candidat au don 2019 (avec age)', dtype={'ÉLIGIBILITÉ_AU_DON.': str})

In [18]:
#dftest['ÉLIGIBILITÉ_AU_DON.']

In [19]:
# ensuring consistency within the columns
clean_df1['Nationalité_'] = clean_df1['Nationalité_'].replace({
    'R A S': 'Camerounaise',
    'Ras': 'Camerounaise'
})
clean_df1['Profession_'] = clean_df1['Profession_'].replace({
    'R A S': 'Pas Précisé',
    'Ras': 'Pas Précisé'
})
clean_df1['Arrondissement_de_résidence_'] = clean_df1['Arrondissement_de_résidence_'].replace({
    'R A S': 'Pas Précisé',
    'Ras': 'Pas Précisé'
})
clean_df1['Quartier_de_Résidence_'] = clean_df1['Quartier_de_Résidence_'].replace({
    'R A S': 'Pas Précisé',
    ' R A S': 'Pas Précisé',
    'Ras': 'Pas Précisé'
})



In [20]:
# Handle Niveau_scolaire column by mapping categories to numerical values
level_mapping = {
    'Pas Précisé': -1,
    'Aucun':0,
    'Primaire': 1,
    'Secondaire': 2,
    'Universitaire': 3,
}

clean_df1['Niveau_scolaire'] = clean_df1['Niveau_scolaire'].replace(level_mapping)

# Check the changes
print(clean_df1[['Religion_', 'Nationalité_', 'Profession_', 'Niveau_scolaire']].head(20))

                 Religion_  Nationalité_       Profession_ Niveau_scolaire
0    Chretien (Catholique)  Camerounaise      Etudiant (e)               3
1    Chretien (Catholique)  Camerounaise      Sans Emplois               3
2              Pas Précisé  Camerounaise         Tailleur                0
3                 Musulman  Camerounaise          Etudiant               3
4               Adventiste  Camerounaise         Militaire               2
5                 Musulman  Camerounaise      Etudiant (e)               3
6              Pas Précisé  Camerounaise      Sans Emplois               1
7                 Musulman  Camerounaise         Bijoutier              -1
8                 Musulman  Camerounaise         Couturier               2
9                 Musulman  Camerounaise       Jeune cadre               3
10                Musulman  Camerounaise    Commerçant (e)               3
11             Pas Précisé  Camerounaise       Mototaximan               2
12                Musulma

C:\Users\hello\AppData\Local\Temp\ipykernel_4684\1320290316.py:10: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  clean_df1['Niveau_scolaire'] = clean_df1['Niveau_scolaire'].replace(level_mapping)
C:\Users\hello\AppData\Local\Temp\ipykernel_4684\1320290316.py:10: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.
  clean_df1['Niveau_scolaire'] = clean_df1['Niveau_scolaire'].replace(level_mapping)


In [21]:
clean_df1.head()

ID  Age Horodateur Niveau_scolaire Genre_  \
0  DONOR_1842   24    46:44.9               3  Homme   
1  DONOR_1843   23    44:19.0               3  Homme   
2  DONOR_1844   41    46:54.3               0  Homme   
3  DONOR_1845   24    29:29.7               3  Homme   
4  DONOR_1846   30    47:47.4               2  Homme   

  Situation_Matrimoniale_(SM)   Profession_ Arrondissement_de_résidence_  \
0                 Célibataire  Etudiant (e)                     Douala 3   
1                 Célibataire  Sans Emplois                  Pas Précisé   
2                   Marié (e)     Tailleur                      Douala 1   
3                 Célibataire      Etudiant                     Douala 5   
4                 Célibataire     Militaire        Douala (Non précisé )   

  Quartier_de_Résidence_  Nationalité_              Religion_  \
0                  Yassa  Camerounaise  Chretien (Catholique)   
1            Pas Précisé  Camerounaise  Chretien (Catholique)   
2               NEW BELL  Camerounaise            Pas Précisé   
3           Ange raphael  Camerounaise               Musulman   
4                  Beedi  Camerounaise             Adventiste   

  A-t-il_(elle)_déjà_donné_le_sang_  Taux_d’hémoglobine_  \
0                               Non            13.895459   
1                               Oui            10.800000   
2                               Non            13.895459   
3                               Non            15.000000   
4                               Oui            13.895459   

           ÉLIGIBILITÉ_AU_DON.  
0                     Eligible  
1  Définitivement non-eligible  
2  Définitivement non-eligible  
3                     Eligible  
4                     Eligible

In [22]:
# Label Encoding for categorical columns
from sklearn.preprocessing import LabelEncoder

# List of columns to encode
columns_to_encode = [
    'Genre_',
    'Situation_Matrimoniale_(SM)',
    'A-t-il_(elle)_déjà_donné_le_sang_',
    'ÉLIGIBILITÉ_AU_DON.'
]

# Initialize the LabelEncoder
label_encoder = LabelEncoder()

# Apply label encoding to specified columns
for column in columns_to_encode:
    clean_df1[column] = label_encoder.fit_transform(clean_df1[column].astype(str))

# Display the updated DataFrame
print("\nDataFrame after Label Encoding:")
print(clean_df1[columns_to_encode].head())


DataFrame after Label Encoding:
   Genre_  Situation_Matrimoniale_(SM)  A-t-il_(elle)_déjà_donné_le_sang_  \
0       1                            0                                  0   
1       1                            0                                  1   
2       1                            2                                  0   
3       1                            0                                  0   
4       1                            0                                  1   

   ÉLIGIBILITÉ_AU_DON.  
0                    1  
1                    0  
2                    0  
3                    1  
4                    1  


In [23]:
clean_df1

ID  Age Horodateur Niveau_scolaire  Genre_  \
0     DONOR_1842   24    46:44.9               3       1   
1     DONOR_1843   23    44:19.0               3       1   
2     DONOR_1844   41    46:54.3               0       1   
3     DONOR_1845   24    29:29.7               3       1   
4     DONOR_1846   30    47:47.4               2       1   
...          ...  ...        ...             ...     ...   
1841  DONOR_3683   26    31:07.0               2       1   
1842  DONOR_3684   23    29:02.4               2       1   
1843  DONOR_3685   30    33:42.3               3       1   
1844  DONOR_3686   28    02:54.5              -1       0   
1845  DONOR_3687   29    56:38.9              -1       1   

      Situation_Matrimoniale_(SM)           Profession_  \
0                               0          Etudiant (e)   
1                               0          Sans Emplois   
2                               2             Tailleur    
3                               0              Etudiant   
4                               0             Militaire   
...                           ...                   ...   
1841                            0                 Élève   
1842                            0                Élève    
1843                            0  Technicien superieur   
1844                            0           Estheticien   
1845                            0           Électricien   

     Arrondissement_de_résidence_ Quartier_de_Résidence_  Nationalité_  \
0                        Douala 3                  Yassa  Camerounaise   
1                     Pas Précisé            Pas Précisé  Camerounaise   
2                        Douala 1               NEW BELL  Camerounaise   
3                        Douala 5           Ange raphael  Camerounaise   
4           Douala (Non précisé )                  Beedi  Camerounaise   
...                           ...                    ...           ...   
1841                     Douala 3               Ndogbong  Camerounaise   
1842        Douala (Non précisé )                   Pk14  Camerounaise   
1843        Douala (Non précisé )      Cite des palmiers  Camerounaise   
1844        Douala (Non précisé )          Douala douala  Camerounaise   
1845                     Douala 3             PK16 NSAPE  Camerounaise   

                  Religion_  A-t-il_(elle)_déjà_donné_le_sang_  \
0     Chretien (Catholique)                                  0   
1     Chretien (Catholique)                                  1   
2               Pas Précisé                                  0   
3                  Musulman                                  0   
4                Adventiste                                  1   
...                     ...                                ...   
1841   Chrétien non précisé                                  0   
1842            Pas Précisé                                  1   
1843               Musulman                                  1   
1844            Pas Précisé                                  0   
1845  Chretien (Catholique)                                  0   

      Taux_d’hémoglobine_  ÉLIGIBILITÉ_AU_DON.  
0               13.895459                    1  
1               10.800000                    0  
2               13.895459                    0  
3               15.000000                    1  
4               13.895459                    1  
...                   ...                  ...  
1841            13.895459                    1  
1842            13.895459                    1  
1843            13.400000                    1  
1844            12.000000                    1  
1845            14.500000                    1  

[1846 rows x 14 columns]

Explaining the numerical values of each column and their category:
- for niveau scolaire :0 represents Aucun, 1 represents Primaire (primary), 2 Secondaire (secondary) , 3 Universitaire (tetiary) and -1 represents Pas Précisé (not specified)
- for Genre_ : 1 for Homme (male) and 0 for Femme (female)
- for Situation_Matrimoniale_(SM): 0 represents celibataire (single), 1 represent divorce (divorced) and 2 represents Marie(married)
- forA-t-il_(elle)_déjà_donné_le_sang_(has he/she donated blood before) 0 represents Non and 1 represents Oui
- for ÉLIGIBILITÉ_AU_DON.: 0 represents Définitivement non-eligible (definitely ineligible), 1 represents eligible , and 2 represents Temporairement Non-eligible(temp. ineligible)

---

### Cleaning 'Donneur 2019' sheet

In [24]:
df2. head(10)

Horodateur Sexe Age  Type de donation   \
0 2020-07-16 19:02:33.994    M   26                 F   
1 2020-07-16 19:05:51.521    M   24                 F   
2 2020-07-17 04:52:09.054    M   41                 F   
3 2020-07-17 04:56:03.200    M   20                 F   
4 2020-07-17 05:49:40.730    M   34                 F   
5 2020-07-20 09:26:29.676    M   34                 F   
6 2020-07-20 09:29:37.199    M   46                 F   
7 2020-07-20 09:34:02.835    M   19                 F   
8 2020-07-20 09:51:21.733    M   40                 F   
9 2020-07-20 09:54:25.363    M   28                 F   

  Groupe Sanguin ABO / Rhesus                       Phenotype   
0                           B+  RH-2, RH+4, RH-3, RH+5, -kell1  
1                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
2                           A+  RH+2, RH+4, RH-3, RH+5, -kell1  
3                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
4                           O+  RH-2, RH+4, RH+3, RH+5, -kell1  
5                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
6                           O+  RH-2, RH+4, RH-3, RH+5, -kell1  
7                           B+  RH-2, RH+2, RH-3, RH+5, -kell1  
8                           O+  RH-2, RH+4, RH+3, RH+5, -kell1  
9                          AB+  RH-2, RH+4, RH-3, RH+5, -kell1

In [25]:
# Display the column names
print(df2.columns.tolist())

['Horodateur', 'Sexe', 'Age ', 'Type de donation ', 'Groupe Sanguin ABO / Rhesus ', 'Phenotype ']


In [26]:
# Count NaN values per column
nan_count = df2.isna().sum()
print(nan_count)

Horodateur                        0
Sexe                              0
Age                               0
Type de donation                221
Groupe Sanguin ABO / Rhesus       0
Phenotype                         0
dtype: int64


In [27]:
df2.shape

(1900, 6)

In [28]:
unique_classes = df2['Type de donation '].unique()
print(unique_classes)

['F' nan 'B']


In the context of blood donation, the letters "F" and "B" typically refer to the following types:

- **F (Full Blood Donation)**: This type involves the collection of whole blood from a donor, which can later be separated into its components (red blood cells, plasma, platelets).

- **B (Blood Component Donation)**: This refers to the donation of specific blood components rather than whole blood. Common component donations include:
  - **Platelet Donation**
  - **Plasma Donation**


performing **Mode Imputation** since it is simple and effective for skewed distributions. Since "F" represents about 90% of the data, using mode imputation ensures that most missing values are filled with the most common value, which is logical in this context.

In [29]:
# Impute missing values with the mode
mode_value = df2['Type de donation '].mode()[0]  # Get the mode
df2['Type de donation ']=  df2['Type de donation '].fillna(mode_value)

# Verify the changes
print(df2['Type de donation '].value_counts())

Type de donation 
F    1845
B      55
Name: count, dtype: int64


In [30]:
# Convert columns to categorical
df2['Type de donation '] = df2['Type de donation '].astype('category')
df2['Groupe Sanguin ABO / Rhesus '] = df2['Groupe Sanguin ABO / Rhesus '].astype('category')
df2['Phenotype '] = df2['Phenotype '].astype('category')

In [31]:
print(df1['Age'].unique())

[24 23 41 30 28 34 31 22 32 27 21 29 26 37 45 35 49 51 46 33 25 19 44  0
 36 20 40 42 38 43 39 48 52 55 47 62 50 56 53 57  1 18 54 16 60 59 58 17
 13 63 12]


In [32]:
# Convert Age to numeric, handling errors
df2['Age '] = pd.to_numeric(df2['Age '], errors='coerce')

In [33]:
# Check for any remaining non-numeric values
print(df1['Age'].isnull().sum())

0


In [34]:
# Check for NaN values
nan_count = df2['Age '].isnull().sum()
print(f"There are {nan_count} NaN values in Age.")

There are 2 NaN values in Age.


In [35]:
df2['Age '] = df2['Age '].fillna(df2['Age '].median())

# convert Age to integer
df2['Age '] = df2['Age '].astype(int)

In [36]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df2['Sexe'] = le.fit_transform(df2['Sexe'])

In [37]:
df2.dtypes

Horodateur                      datetime64[ns]
Sexe                                     int32
Age                                      int32
Type de donation                      category
Groupe Sanguin ABO / Rhesus           category
Phenotype                             category
dtype: object

In [38]:
# Extract the date and ensure it's a separate copy
df2['Date'] = df2['Horodateur'].dt.date.copy() 

# Now drop the 'horadateur' column
df2.drop(columns=['Horodateur'], inplace=True)

In [39]:
clean_df2 = df2
clean_df2

Sexe  Age  Type de donation  Groupe Sanguin ABO / Rhesus   \
0        1    26                 F                           B+   
1        1    24                 F                           O+   
2        1    41                 F                           A+   
3        1    20                 F                           O+   
4        1    34                 F                           O+   
...    ...   ...               ...                          ...   
1895     1    29                 F                           O-   
1896     1    35                 F                           O-   
1897     1    21                 F                           A+   
1898     0    19                 F                           O+   
1899     1    32                 F                           O+   

                          Phenotype         Date  
0     RH-2, RH+4, RH-3, RH+5, -kell1  2020-07-16  
1     RH-2, RH+4, RH-3, RH+5, -kell1  2020-07-16  
2     RH+2, RH+4, RH-3, RH+5, -kell1  2020-07-17  
3     RH-2, RH+4, RH-3, RH+5, -kell1  2020-07-17  
4     RH-2, RH+4, RH+3, RH+5, -kell1  2020-07-17  
...                              ...         ...  
1895  RH-2, RH+4, RH-3, RH+5, -kell1  2020-07-24  
1896  RH-2, RH+4, RH-3, RH+5, -kell1  2020-07-24  
1897  RH-2, RH+4, RH-3, RH+5, -kell1  2020-07-24  
1898  RH+2, RH+4, RH-3, RH+5, -kell1  2020-07-24  
1899  RH-2, RH+4, RH+3, RH+5, -kell1  2020-07-24  

[1900 rows x 6 columns]

In [40]:
file_name = 'final_clean_data.xlsx'

# Use ExcelWriter to save multiple DataFrames to different sheets
with pd.ExcelWriter(file_name) as writer:
    clean_df1.to_excel(writer, sheet_name='Clean_Candidate_Donor_data', index=False)
    clean_df2.to_excel(writer, sheet_name='Clean_Donor_Data', index=False)